File: hapsigntool/hap_sign_tool/src/main/java/com/ohos/hapsigntool/HapSignTool.java
/*
 * Copyright (c) 2021-2022 Huawei Device Co., Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.ohos.hapsigntool;


import com.ohos.hapsigntool.api.ServiceApi;
import com.ohos.hapsigntool.api.SignToolServiceImpl;
import com.ohos.hapsigntool.api.model.Options;
import com.ohos.hapsigntool.error.CustomException;
import com.ohos.hapsigntool.error.ERROR;
import com.ohos.hapsigntool.utils.FileUtils;
import com.ohos.hapsigntool.utils.StringUtils;
import com.ohos.hapsigntoolcmd.CmdUtil;
import com.ohos.hapsigntoolcmd.CmdUtil.Method;
import com.ohos.hapsigntoolcmd.HelpDocument;
import com.ohos.hapsigntoolcmd.Params;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

/**
 * HapSignTool.
 *
 * @since 2021/12/28
 */
public final class HapSignTool {
    /**
     * Add log info.
     */
    private static final Logger LOGGER = LogManager.getLogger(HapSignTool.class);

    /**
     * Tool version.
     */
    private static final String VERSION = "1.0.0";

    /**
     * Local sign.
     */
    private static final String LOCAL_SIGN = "localSign";

    /**
     * Remote sign.
     */
    private static final String REMOTE_SIGN = "remoteSign";

    /**
     * Signed.
     */
    private static final String SIGNED = "1";

    /**
     * No signed.
     */
    private static final String NOT_SIGNED = "0";

    private HapSignTool() {
    }

    /**
     * Main entry.
     *
     * @param args arguments
     */
    public static void main(String[] args) {
        try {
            boolean result = processCmd(args);
            if (!result) {
                System.exit(1);
            }
        } catch (CustomException exception) {
            LOGGER.debug(exception.getMessage(), exception);
            LOGGER.error(exception.getMessage());
        }
    }

    /**
     * Process command.
     *
     * @param args arguments
     * @return command processing result
     * @throws CustomException custom exception for command execution failure
     */
    public static boolean processCmd(String[] args) throws CustomException {
        if (args.length == 0 || StringUtils.isEmpty(args[0])) {
            help();
        } else if ("-h".equals(args[0]) || "-help".equals(args[0])) {
            help();
        } else if ("-v".equals(args[0]) || "-version".equals(args[0])) {
            version();
        } else {
            ServiceApi api = new SignToolServiceImpl();
            Params params = CmdUtil.convert2Params(args);
            LOGGER.debug(params.toString());
            LOGGER.info("Start {}", params.getMethod());
            boolean result;
            result = dispatchParams(params, api);
            if (result) {
                LOGGER.info(String.format("%s %s", params.getMethod(), "success"));
            } else {
                LOGGER.info(String.format("%s %s", params.getMethod(), "failed"));
            }
            return result;
        }
        return true;
    }

    private static boolean callGenerators(Params params, ServiceApi api) {
        boolean result = false;
        switch (params.getMethod()) {
            case Method.GENERATE_APP_CERT:
                result = runAppCert(params.getOptions(), api);
                break;
            case Method.GENERATE_CA:
                result = runCa(params.getOptions(), api);
                break;
            case Method.GENERATE_CERT:
                result = runCert(params.getOptions(), api);
                break;
            case Method.GENERATE_CSR:
                result = runCsr(params.getOptions(), api);
                break;
            case Method.GENERATE_KEYPAIR:
                result = runKeypair(params.getOptions(), api);
                break;
            case Method.GENERATE_PROFILE_CERT:
                result = runProfileCert(params.getOptions(), api);
                break;
            default:
                CustomException.throwException(ERROR.COMMAND_ERROR, "Unsupported cmd");
                break;
        }
        return result;
    }

    private static boolean dispatchParams(Params params, ServiceApi api) {
        boolean result;
        switch (params.getMethod()) {
            case Method.SIGN_APP:
                result = runSignApp(params.getOptions(), api);
                break;
            case Method.SIGN_PROFILE:
                result = runSignProfile(params.getOptions(), api);
                break;
            case Method.VERIFY_APP:
                result = runVerifyApp(params.getOptions(), api);
                break;
            case Method.VERIFY_PROFILE:
                result = runVerifyProfile(params.getOptions(), api);
                break;
            default:
                result = callGenerators(params, api);
                break;
        }

        return result;
    }

    private static void checkEndCertArguments(Options params) {
        params.required(Options.KEY_ALIAS, Options.ISSUER, Options.ISSUER_KEY_ALIAS, Options.SUBJECT,
                Options.SIGN_ALG, Options.KEY_STORE_FILE);
        String signAlg = params.getString(Options.SIGN_ALG);
        CmdUtil.judgeEndSignAlgType(signAlg);
        String outForm = params.getString(Options.OUT_FORM);
        if (!StringUtils.isEmpty(outForm)) {
            CmdUtil.verifyType(outForm, Options.OUT_FORM_SCOPE);
        }
        if (StringUtils.isEmpty(outForm) || "certChain".equals(outForm)) {
            params.required(Options.SUB_CA_CERT_FILE, Options.CA_CERT_FILE);
            FileUtils.validFileType(params.getString(Options.SUB_CA_CERT_FILE), "cer");
            FileUtils.validFileType(params.getString(Options.CA_CERT_FILE), "cer");
        }
        String keyStoreFile = params.getString(Options.KEY_STORE_FILE);
        FileUtils.validFileType(keyStoreFile, "p12", "jks");

        if (params.containsKey(Options.ISSUER_KEY_STORE_FILE)){
            String issuerKeyStoreFile = params.getString(Options.ISSUER_KEY_STORE_FILE);
            FileUtils.validFileType(issuerKeyStoreFile, "p12", "jks");
        }

        String outFile = params.getString(Options.OUT_FILE);
        if (!StringUtils.isEmpty(outFile)) {
            FileUtils.validFileType(outFile, "cer", "pem");
        }
    }

    private static boolean runAppCert(Options params, ServiceApi api) {
        checkEndCertArguments(params);
        return api.generateAppCert(params);
    }

    private static boolean runCa(Options params, ServiceApi api) {
        params.required(Options.KEY_ALIAS, Options.KEY_ALG, Options.KEY_SIZE, Options.SUBJECT,
                Options.SIGN_ALG, Options.KEY_STORE_FILE);
        String keyAlg = params.getString(Options.KEY_ALG);
        CmdUtil.judgeAlgType(keyAlg);
        String size = params.getString(Options.KEY_SIZE);
        CmdUtil.judgeSize(size, keyAlg);
        String signAlg = params.getString(Options.SIGN_ALG);
        CmdUtil.judgeSignAlgType(signAlg);
        FileUtils.validFileType(params.getString(Options.KEY_STORE_FILE), "p12", "jks");
        params.put(Options.KEY_SIZE, CmdUtil.convertAlgSize(size));
        return api.generateCA(params);
    }

    private static boolean runCert(Options params, ServiceApi api) {
        params.required(Options.KEY_ALIAS, Options.ISSUER, Options.ISSUER_KEY_ALIAS, Options.SUBJECT,
                Options.KEY_USAGE, Options.SIGN_ALG, Options.KEY_STORE_FILE);
        String keyUsage = params.getString(Options.KEY_USAGE);
        CmdUtil.verifyType(keyUsage, Options.KEY_USAGE_SCOPE);
        String extKeyUsage = params.getString(Options.EXT_KEY_USAGE);
        CmdUtil.verifyType(extKeyUsage, Options.EXT_KEY_USAGE_SCOPE);
        String signAlg = params.getString(Options.SIGN_ALG);
        CmdUtil.judgeSignAlgType(signAlg);
        FileUtils.validFileType(params.getString(Options.KEY_STORE_FILE), "p12", "jks");
        if (params.containsKey(Options.ISSUER_KEY_STORE_FILE)){
            String issuerKeyStoreFile = params.getString(Options.ISSUER_KEY_STORE_FILE);
            FileUtils.validFileType(issuerKeyStoreFile, "p12", "jks");
        }
        return api.generateCert(params);
    }

    private static boolean runCsr(Options params, ServiceApi api) {
        params.required(Options.KEY_ALIAS, Options.SUBJECT, Options.SIGN_ALG, Options.KEY_STORE_FILE);
        String signAlg = params.getString(Options.SIGN_ALG);
        CmdUtil.judgeSignAlgType(signAlg);
        FileUtils.validFileType(params.getString(Options.KEY_STORE_FILE), "p12", "jks");
        if (!StringUtils.isEmpty(params.getString(Options.OUT_FILE))) {
            FileUtils.validFileType(params.getString(Options.OUT_FILE), "csr");
        }

        return api.generateCsr(params);
    }

    private static boolean runKeypair(Options params, ServiceApi api) {
        params.required(Options.KEY_ALIAS, Options.KEY_ALG, Options.KEY_SIZE, Options.KEY_STORE_FILE);
        String keyAlg = params.getString(Options.KEY_ALG);
        CmdUtil.judgeAlgType(keyAlg);
        String size = params.getString(Options.KEY_SIZE);
        CmdUtil.judgeSize(size, keyAlg);
        params.put(Options.KEY_SIZE, CmdUtil.convertAlgSize(size));
        FileUtils.validFileType(params.getString(Options.KEY_STORE_FILE), "p12", "jks");

        return api.generateKeyStore(params);
    }

    private static boolean runProfileCert(Options params, ServiceApi api) {
        checkEndCertArguments(params);
        return api.generateProfileCert(params);
    }

    private static boolean runSignApp(Options params, ServiceApi api) {
        params.required(Options.MODE, Options.IN_FILE, Options.OUT_FILE, Options.PROFILE_FILE, Options.SIGN_ALG);
        String mode = params.getString(Options.MODE);
        if (!LOCAL_SIGN.equalsIgnoreCase(mode)
                && !REMOTE_SIGN.equalsIgnoreCase(mode)
                && !"remoteResign".equalsIgnoreCase(mode)) {
            CustomException.throwException(ERROR.COMMAND_ERROR, "mode params is incorrect");
        }

        if (LOCAL_SIGN.equalsIgnoreCase(mode)) {
            params.required(Options.KEY_STORE_FILE, Options.KEY_ALIAS, Options.APP_CERT_FILE);
            FileUtils.validFileType(params.getString(Options.KEY_STORE_FILE), "p12", "jks");
        }
        checkProfile(params);
        String inForm = params.getString(Options.IN_FORM);
        if (!StringUtils.isEmpty(inForm) && !"zip".equalsIgnoreCase(inForm) && !"bin".equalsIgnoreCase(inForm)) {
            CustomException.throwException(ERROR.NOT_SUPPORT_ERROR, "inForm params is incorrect");
        }
        String signAlg = params.getString(Options.SIGN_ALG);
        CmdUtil.judgeEndSignAlgType(signAlg);

        return api.signHap(params);
    }

    private static void checkProfile(Options params) {
        String profileFile = params.getString(Options.PROFILE_FILE);
        String profileSigned = params.getString(Options.PROFILE_SIGNED,SIGNED);
        if (!SIGNED.equals(profileSigned) && !NOT_SIGNED.equals(profileSigned)) {
            CustomException.throwException(ERROR.NOT_SUPPORT_ERROR, "profileSigned params is incorrect");
        }
        if (SIGNED.equals(profileSigned)) {
            FileUtils.validFileType(profileFile, "p7b");
        } else {
            FileUtils.validFileType(profileFile, "json");
        }
    }

    private static boolean runSignProfile(Options params, ServiceApi api) {
        params.required(Options.MODE, Options.SIGN_ALG, Options.OUT_FILE, Options.IN_FILE);
        String mode = params.getString(Options.MODE);
        if (!LOCAL_SIGN.equalsIgnoreCase(mode) && !REMOTE_SIGN.equalsIgnoreCase(mode)) {
            CustomException.throwException(ERROR.COMMAND_ERROR, "mode params is incorrect");
        }
        if (LOCAL_SIGN.equalsIgnoreCase(mode)) {
            params.required(Options.KEY_STORE_FILE, Options.KEY_ALIAS, Options.PROFILE_CERT_FILE);
            FileUtils.validFileType(params.getString(Options.KEY_STORE_FILE), "p12", "jks");
        }

        String signAlg = params.getString(Options.SIGN_ALG);
        CmdUtil.judgeEndSignAlgType(signAlg);
        String outFile = params.getString(Options.OUT_FILE);
        FileUtils.validFileType(outFile, "p7b");

        return api.signProfile(params);
    }

    private static boolean runVerifyApp(Options params, ServiceApi api) {
        params.required(Options.IN_FILE, Options.OUT_CERT_CHAIN,
                Options.OUT_PROFILE);
        FileUtils.validFileType(params.getString(Options.IN_FILE), "hap", "bin");
        FileUtils.validFileType(params.getString(Options.OUT_CERT_CHAIN), "cer");
        FileUtils.validFileType(params.getString(Options.OUT_PROFILE), "p7b");
        return api.verifyHap(params);
    }

    private static boolean runVerifyProfile(Options params, ServiceApi api) {
        params.required(Options.IN_FILE);
        FileUtils.validFileType(params.getString(Options.IN_FILE), "p7b");
        String outFile = params.getString(Options.OUT_FILE);
        if (!StringUtils.isEmpty(outFile)) {
            FileUtils.validFileType(outFile, "json");
        }

        return api.verifyProfile(params);
    }

    /**
     * Software version.
     */
    public static void version() {
        LOGGER.info(VERSION);
    }

    /**
     * Print help to console.
     */
    public static void help() {
        HelpDocument.printHelp(LOGGER);
    }
}

File: pc/services/cmd/src/hap_sign_tool.cpp
/*
 * Copyright (c) Huawei Technologies Co., Ltd. 2024-2024. All rights reserved.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
#include "hap_sign_tool.h"
#include "constant.h"
#include <memory>

using namespace OHOS::SignatureTools;
using std::pair;
const std::string HapSignTool::VERSION = "1.0.0";
const std::string HapSignTool::LOCAL_SIGN = "localSign";
const std::string HapSignTool::REMOTE_SIGN = "remoteSign";

std::vector<std::string> HapSignTool::InformList;
enum MapNum {
    NUM_1 = 1,
    NUM_2 = 2,
    NUM_3 = 3,
    NUM_4 = 4,
    NUM_5 = 5,
    NUM_6 = 6
};
HapSignTool::HapSignTool()
{
    InformList.push_back("bin");
    InformList.push_back("elf");
    InformList.push_back("zip");
}
bool HapSignTool::ProcessCmd(char** args, size_t size)
{
    if (size < CmdUtil::ARGS_MIN_LEN) {
        args[1] = const_cast<char*>("");
    }
    if (args == nullptr || strcmp(args[1], "") == 0) {
        PrintHelp();
        return true;
    } else if (strcmp(args[1], "-h") == 0 || strcmp(args[1], "-help") == 0) {
        PrintHelp();
        return true;
    } else if (strcmp(args[1], "-v") == 0 || strcmp(args[1], "-version") == 0) {
        Version();
        return true;
    } else {
        std::shared_ptr<SignToolServiceImpl> service_api = std::make_shared<SignToolServiceImpl>();
        CmdUtil cmdUtil;
        ParamsSharedPtr param = std::make_shared<Params>();
        if (!cmdUtil.Convert2Params(args, size, param)) {
            return false;
        }
        CMD_MSG("Start " + param->GetMethod());
        SIGNATURE_TOOLS_LOGI("%{public}s run start time \n ", param->GetMethod().c_str());
        if (!DispatchParams(param, *service_api.get())) {
            SIGNATURE_TOOLS_LOGI("%{public}s run end time \n ", param->GetMethod().c_str());
            return false;
        }
        CMD_MSG(param->GetMethod() + " success");
        SIGNATURE_TOOLS_LOGI("%{public}s run end time \n ", param->GetMethod().c_str());
    }
    return true;
}
bool HapSignTool::CallGenerators(ParamsSharedPtr params, SignToolServiceImpl& api)
{
    bool isSuccess = false;
    std::map<std::string, int> map;
    map.insert(std::make_pair(Method::GENERATE_KEYPAIR, MapNum::NUM_1));
    map.insert(std::make_pair(Method::GENERATE_CSR, MapNum::NUM_2));
    map.insert(std::make_pair(Method::GENERATE_CA, MapNum::NUM_3));
    map.insert(std::make_pair(Method::GENERATE_APP_CERT, MapNum::NUM_4));
    map.insert(std::make_pair(Method::GENERATE_PROFILE_CERT, MapNum::NUM_5));
    map.insert(std::make_pair(Method::GENERATE_CERT, MapNum::NUM_6));
    switch (map[params->GetMethod()]) {
        case MapNum::NUM_1:
            isSuccess = RunKeypair(params->GetOptions(), api);
            break;
        case MapNum::NUM_2:
            isSuccess = RunCsr(params->GetOptions(), api);
            break;
        case MapNum::NUM_3:
            isSuccess = RunCa(params->GetOptions(), api);
            break;
        case MapNum::NUM_4:
            isSuccess = RunAppCert(params->GetOptions(), api);
            break;
        case MapNum::NUM_5:
            isSuccess = RunProfileCert(params->GetOptions(), api);
            break;
        case MapNum::NUM_6:
            isSuccess = RunCert(params->GetOptions(), api);
            break;
        default:
            break;
    }
    return isSuccess;
}
bool HapSignTool::RunSignApp(Options* params, SignToolServiceImpl& api)
{
    if (!params->Required({ Options::MODE, Options::IN_FILE, Options::OUT_FILE, Options::SIGN_ALG })) {
        return false;
    }
    std::string mode = params->GetString(Options::MODE);
    std::string remote = "remoteResign";
    if (!StringUtils::CaseCompare(mode, HapSignTool::LOCAL_SIGN) &&
        !StringUtils::CaseCompare(mode, HapSignTool::REMOTE_SIGN) &&
        !StringUtils::CaseCompare(mode, remote)) {
        CMD_ERROR_MSG("COMMAND_ERROR", COMMAND_ERROR, "mode params is incorrect");
        return false;
    }
    if (StringUtils::CaseCompare(mode, HapSignTool::LOCAL_SIGN)) {
        if (!params->Required({ Options::KEY_STORE_FILE, Options::KEY_ALIAS, Options::APP_CERT_FILE })) {
            return false;
        }
        if (!FileUtils::ValidFileType(params->GetString(Options::KEY_STORE_FILE), { "p12", "jks" })) {
            return false;
        }
    }
    if (!CheckProfile(*params)) {
        return false;
    }
    std::string zip = "zip";
    std::string inForm = params->GetString(Options::INFORM, zip);
    if (!StringUtils::IsEmpty(inForm) && !StringUtils::ContainsCase(InformList, inForm)) {
        CMD_ERROR_MSG("NOT_SUPPORT_ERROR", NOT_SUPPORT_ERROR, "inForm params is incorrect");
        return false;
    }
    std::string signAlg = params->GetString(Options::SIGN_ALG);
    if (!CmdUtil::JudgeEndSignAlgType(signAlg)) {
        return false;
    }
    return api.SignHap(params);
}
bool HapSignTool::CheckProfile(Options& params)
{
    std::string inForm = params.GetString(Options::INFORM);
    std::string profileFile = params.GetString(Options::PROFILE_FILE);

    std::string profileSigned = params.GetString(Options::PROFILE_SIGNED);
    std::string elf = "elf";
    if (StringUtils::CaseCompare(inForm, elf) && FileUtils::IsEmpty(profileFile)) {
        SIGNATURE_TOOLS_LOGW("INFORM with PROFILE_FILE Check error ! \n");
        return true;
    }

    if (profileSigned == "1") {
        if (!FileUtils::ValidFileType(profileFile, { "p7b" })) {
            return false;
        }
        return true;
    } else {
        if (!FileUtils::ValidFileType(profileFile, { "json" })) {
            return false;
        }
        return true;
    }
}
bool HapSignTool::DispatchParams(ParamsSharedPtr params, SignToolServiceImpl& api)
{
    bool isSuccess = false;
    std::map<std::string, int> map;
    map.insert(std::make_pair(Method::SIGN_APP, MapNum::NUM_1));
    map.insert(std::make_pair(Method::SIGN_PROFILE, MapNum::NUM_2));
    map.insert(std::make_pair(Method::VERIFY_APP, MapNum::NUM_3));
    map.insert(std::make_pair(Method::VERIFY_PROFILE, MapNum::NUM_4));
    switch (map[params->GetMethod()]) {
        case MapNum::NUM_1:
            isSuccess = RunSignApp(params->GetOptions(), api);
            break;
        case MapNum::NUM_2:
            isSuccess = HapSignTool::RunSignProfile(params->GetOptions(), api);
            break;
        case MapNum::NUM_3:
            isSuccess = RunVerifyApp(params->GetOptions(), api);
            break;
        case MapNum::NUM_4:
            isSuccess = HapSignTool::RunVerifyProfile(params->GetOptions(), api);
            break;
        default:
            isSuccess = HapSignTool::CallGenerators(params, api);
            break;
    }
    return isSuccess;
}



bool HapSignTool::RunCa(Options* params, SignToolServiceImpl& api)
{
    if (!params->Required({ Options::KEY_ALIAS, Options::KEY_ALG,
        Options::KEY_SIZE, Options::SUBJECT, Options::SIGN_ALG, Options::KEY_STORE_FILE })) {
        return false;
    }
    std::string keyAlg = params->GetString(Options::KEY_ALG);
    if (!CmdUtil::JudgeAlgType(keyAlg)) {
        return false;
    }
    int size = params->GetInt(Options::KEY_SIZE);
    if (!CmdUtil::JudgeSize(size)) {
        return false;
    }
    std::string signAlg = params->GetString(Options::SIGN_ALG);
    if (!CmdUtil::JudgeSignAlgType(signAlg)) {
        return false;
    }
    if (!FileUtils::ValidFileType(params->GetString(Options::KEY_STORE_FILE), { "p12", "jks" })) {
        return false;
    }
    
    return api.GenerateCA(params);
}
bool HapSignTool::RunCert(Options* params, SignToolServiceImpl& api)
{
    if (!params->Required({ Options::KEY_ALIAS, Options::ISSUER,
        Options::ISSUER_KEY_ALIAS, Options::SUBJECT, Options::KEY_USAGE,
        Options::SIGN_ALG, Options::KEY_STORE_FILE })) {
        return false;
    }
    std::string keyusage = params->GetString(Options::KEY_USAGE);
    if (!CmdUtil::VerifyTypes(keyusage)) {
        return false;
    }
    std::string extkeyusage = params->GetString(Options::EXT_KEY_USAGE);
    if (!extkeyusage.empty()) {
        if (!CmdUtil::VerifyType(extkeyusage)) {
            return false;
        }
    }
    std::string signAlg = params->GetString(Options::SIGN_ALG);
    if (!CmdUtil::JudgeSignAlgType(signAlg)) {
        return false;
    }
    if (!FileUtils::ValidFileType(params->GetString(Options::KEY_STORE_FILE), { "p12", "jks" })) {
        return false;
    }
    return api.GenerateCert(params);
}
bool HapSignTool::CheckEndCertArguments(Options& params)
{
    if (!params.Required({ params.KEY_ALIAS, params.ISSUER, params.ISSUER_KEY_ALIAS,
                        params.SUBJECT, params.SIGN_ALG, params.KEY_STORE_FILE })) {
        return false;
    }
    std::string signAlg = params.GetString(params.SIGN_ALG);
    if (!CmdUtil::JudgeSignAlgType(signAlg)) {
        return false;
    }
    std::string outForm = params.GetString(params.OUT_FORM);
    if (!outForm.empty()) {
        if (!CmdUtil::VerifyType(outForm, Options::OUT_FORM_SCOPE)) {
        }
    }
    if (!outForm.empty() && "certChain" == outForm) {
        if (!params.Required({ params.SUB_CA_CERT_FILE, params.CA_CERT_FILE })) {
            return false;
        }
        if (!FileUtils::ValidFileType(params.GetString(params.SUB_CA_CERT_FILE), { "cer" })
            || !FileUtils::ValidFileType(params.GetString(params.CA_CERT_FILE), { "cer" })) {
            return false;
        }
    }
    std::string keyStoreFile = params.GetString(params.KEY_STORE_FILE);
    if (!FileUtils::ValidFileType(keyStoreFile, { "p12", "jks" })) {
        return false;
    }
    if (params.find(params.ISSUER_KEY_STORE_FILE) != params.end()) {
        std::string issuerKeyStoreFile = params.GetString(params.ISSUER_KEY_STORE_FILE);
        if (!FileUtils::ValidFileType(issuerKeyStoreFile, { "p12", "jks" })) {
        }
    }
    std::string outFile = params.GetString(params.OUT_FILE);
    if (!outFile.empty()) {
        FileUtils::ValidFileType(outFile, { "cer", "pem" });
    }
    return true;
}
bool HapSignTool::RunAppCert(Options* params, SignToolServiceImpl& api)
{
    if (!HapSignTool::CheckEndCertArguments(*params)) {
        return false;
    }
    return api.GenerateAppCert(params);
}
bool HapSignTool::RunProfileCert(Options* params, SignToolServiceImpl& api)
{
    if (!HapSignTool::CheckEndCertArguments(*params)) {
        return false;
    }
    return api.GenerateProfileCert(params);
}
bool HapSignTool::RunKeypair(Options* params, SignToolServiceImpl& api)
{
    if (!params->Required({ Options::KEY_ALIAS, Options::KEY_ALG, Options::KEY_SIZE, Options::KEY_STORE_FILE })) {
        return false;
    }
    std::string keyAlg = params->GetString(Options::KEY_ALG);
    if (!CmdUtil::JudgeAlgType(keyAlg)) {
        return false;
    }
    int size = params->GetInt(Options::KEY_SIZE);
    if (!CmdUtil::JudgeSize(size)) {
        return false;
    }
    if (!FileUtils::ValidFileType(params->GetString(Options::KEY_STORE_FILE), { "p12", "jks" })) {
        return false;
    }
    return api.GenerateKeyStore(params);
}
bool HapSignTool::RunCsr(Options* params, SignToolServiceImpl& api)
{
    if (!params->Required({ Options::KEY_ALIAS, Options::SUBJECT, Options::SIGN_ALG, Options::KEY_STORE_FILE })) {
        return false;
    }
    std::string signAlg = params->GetString(Options::SIGN_ALG);
    if (!CmdUtil::JudgeSignAlgType(signAlg)) {
        return false;
    }
    if (!FileUtils::ValidFileType(params->GetString(Options::KEY_STORE_FILE), { "p12", "jks" })) {
        return false;
    }
    if (!FileUtils::ValidFileType(params->GetString(Options::OUT_FILE), { "csr" })) {
    }
    return api.GenerateCsr(params);
}
bool HapSignTool::RunSignProfile(Options* params, SignToolServiceImpl& api)
{
    if (params->Required({ params->MODE, params->SIGN_ALG, params->OUT_FILE, params->IN_FILE }) == false) {
        return false;
    }
    std::string mode = params->GetString(Options::MODE);
    if (!StringUtils::CaseCompare(mode, HapSignTool::LOCAL_SIGN) &&
        !StringUtils::CaseCompare(mode, HapSignTool::REMOTE_SIGN)) {
        CMD_ERROR_MSG("COMMAND_ERROR", COMMAND_ERROR, "mode params is incorrect");
        return false;
    }
    if (StringUtils::CaseCompare(mode, HapSignTool::LOCAL_SIGN)) {
        if (params->Required({ params->KEY_STORE_FILE, params->KEY_ALIAS, params->PROFILE_CERT_FILE }) == false)
            return false;
        if (FileUtils::ValidFileType(params->GetString(Options::KEY_STORE_FILE), { "p12", "jks" }) == false)
            return false;
    }
    std::string signAlg = params->GetString(Options::SIGN_ALG);
    if (CmdUtil::JudgeEndSignAlgType(signAlg) == false) {
        return false;
    }
    std::string outFile = params->GetString(Options::OUT_FILE);
    if (FileUtils::ValidFileType(outFile, { "p7b" }) == false) {
        return false;
    }
    return api.SignProfile(params);
}
bool HapSignTool::RunVerifyProfile(Options* params, SignToolServiceImpl& api)
{
    if (params->Required({ Options::IN_FILE }) == false)
        return false;
    if (FileUtils::ValidFileType(params->GetString(Options::IN_FILE), { "p7b" }) == false)
        return true;
    std::string outFile = params->GetString(Options::OUT_FILE);
    if (!outFile.empty()) {
        if (FileUtils::ValidFileType(outFile, { "json" }) == false)
            return false;
    }
    return api.VerifyProfile(params);
}
void HapSignTool::PrintHelp()
{
    std::ifstream readHelp(HELP_FILE_PATH.c_str(), std::ios::in | std::ios::binary);
    if (readHelp.is_open()) {
        std::string line;
        while (std::getline(readHelp, line)) {
            printf("%s\n", line.c_str());
        }
        readHelp.close();
    } else {
        CMD_ERROR_MSG("OPEN_FILE_ERROR", OPEN_FILE_ERROR, "Open " + HELP_FILE_PATH + " failed");
    }
}
void  HapSignTool::Version()
{
    printf("%s\n", HapSignTool::VERSION.c_str());
}
bool HapSignTool::RunVerifyApp(Options* params, SignToolServiceImpl& api)
{
    if (!params->Required({ Options::IN_FILE, Options::OUT_CERT_CHAIN, Options::OUT_PROFILE })) {
        return false;
    }
    std::string zip = "zip";
    std::string inForm = params->GetString(Options::INFORM, zip);
    if (!StringUtils::ContainsCase(InformList, inForm)) {
        CMD_ERROR_MSG("NOT_SUPPORT_ERROR", NOT_SUPPORT_ERROR, "inForm params must is [bin, elf, zip]");
        return false;
    }
    if (!FileUtils::ValidFileType(params->GetString(Options::OUT_CERT_CHAIN), { "cer" })) {
        return false;
    }
    if (!FileUtils::ValidFileType(params->GetString(Options::OUT_PROFILE), { "p7b" })) {
        return false;
    }
    return api.VerifyHap(params);
}